### PR TITLE
Docker command customization documentation

### PR DESCRIPTION
**Pull request by bwateratmsft**
Adds documentation for the Docker command customization features. The content is adapted from https://github.com/microsoft/vscode-docker/issues/1596#issue-558411927.

@karolz-ms, @philliphoff, @BigMorty for FYI.

This takes care of https://github.com/microsoft/vscode-docker/issues/1653.

### CODE REVIEW

**Comment by bwateratmsft** [comment]
@ghogen @gregvanl can you review? Thank you!

File: docs/containers/reference.md [hunk 1]
```diff
@@ -268,3 +268,147 @@ Here are all properties available for configuring `docker-run` task. All propert
...
+
+Supported tokens:
+
+| Token | Description |
+| -- | -- |
+| `${dockerfile}` | The workspace-relative path of the selected `Dockerfile`. |
+| `${tag}` | The value entered/confirmed by the user upon invoking the build command. If previously built, defaults to the previously-entered value for that `Dockerfile`. |
+| `${context}` | If set, the value of the `docker.imageBuildContextPath` configuration setting. Otherwise, the workspace-relative folder in which the `Dockerfile` resides. |
+
+> NOTE: If the `docker.commands.build` setting does not contain the `${tag}` token, the user will *not* be prompted to enter/confirm a tag.
```
Comment (ghogen) [inline docs/containers/reference.md]: Matching the style used in the VS Code docs for notes.
```suggestion
> **Note**: If the `docker.commands.build` setting does not contain the `${tag}` token, the user will *not* be prompted to enter/confirm a tag.
```

File: docs/containers/reference.md [hunk 2]
```diff
@@ -268,3 +268,147 @@ Here are all properties available for configuring `docker-run` task. All propert
...
+| `docker.commands.runInteractive` | `docker run --rm -it ${exposedPorts} ${tag}` |
+
+Supported tokens:
+
+| Token | Description |
+| -- | -- |
+| `${exposedPorts}` | Generated from the list of exposed ports in the image (i.e. ultimately from the `Dockerfile`), where each exposed port is mapped to the same port on the local machine.  For example, `"EXPOSE 5000 5001"` would generate `"-p 5000:5000 -p 5001:5001"`. |
+| `${tag}` | The full tag of the selected image. |
+
+> NOTE: The `match` regular expression will be compared against the full tag of the selected image.
```
Comment (ghogen) [inline docs/containers/reference.md]:
```suggestion
> **Note**: The `match` regular expression will be compared against the full tag of the selected image.
```

File: docs/containers/reference.md [hunk 3]
```diff
@@ -268,3 +268,147 @@ Here are all properties available for configuring `docker-run` task. All propert
...
+| `docker.commands.attach` | `docker exec -it ${containerId} ${shellCommand}`
+
+Supported tokens:
+
+| Token | Description |
+| -- | -- |
+| `${containerId}` | The ID of the container to attach to. |
+| `${shellCommand}` | The value of the `docker.attachShellCommand.linuxContainer` or `docker.attachShellCommand.windowsContainer` configuration setting, as appropriate. |
+
+> NOTE: The `match` regular expression will be compared against the container name and full tag of the container image.
```
Comment (ghogen) [inline docs/containers/reference.md]:
```suggestion
> **Note**: The `match` regular expression will be compared against the container name and full tag of the container image.
```

File: docs/containers/reference.md [hunk 4]
```diff
@@ -268,3 +268,147 @@ Here are all properties available for configuring `docker-run` task. All propert
...
+|--|--|
+| `docker.commands.logs` | `docker logs -f ${containerId}`
+
+Supported tokens:
+
+| Token | Description |
+| -- | -- |
+| `${containerId}` | The ID of the container to view the logs for. |
+
+> NOTE: The `match` regular expression will be compared against the container name and full tag of the container image.
```
Comment (ghogen) [inline docs/containers/reference.md]:
```suggestion
> **Note**: The `match` regular expression will be compared against the container name and full tag of the container image.
```

File: docs/containers/reference.md [hunk 5]
```diff
@@ -268,3 +268,147 @@ Here are all properties available for configuring `docker-run` task. All propert
 | `enableDebugging` |  Whether or not to enable debugging within the container. | `false` |
 | `inspectMode` |  Defines the initial interaction between the application and the debugger (`default` or `break`). <br/> The value `default` allows the application to run until the debugger attaches. <br/> The value `break` prevents the application from running until the debugger attaches. | `default` |
 | `inspectPort` |  The port on which debugging should occur. | `9229` |
+
+## Command customization
+
+The Docker extension executes a number of Docker CLI commands on behalf of its users, such as to build images, run containers, attach to containers, and view container logs. Some of these commands have a large number of optional arguments, often used in very specific scenarios. Many of these commands can be customized.
```
Comment (ghogen) [inline docs/containers/reference.md]: "on behalf of its users" I think can be improved.  Can this be more exact?  Something like "when you perform various operations, such as to build images, run containers ..."

File: docs/containers/reference.md [hunk 6]
```diff
@@ -268,3 +268,147 @@ Here are all properties available for configuring `docker-run` task. All propert
...
+
+For each of these customizable Docker commands, a configuration setting is available to set the template of what to execute. Alternatively, multiple templates can be defined, optionally with a regular expression which, when matched, hints the context in which a template should be used. The templates support some tokens similar to `launch.json` and `tasks.json`, for example, `${workspaceFolder}`.
```
Comment (ghogen) [inline docs/containers/reference.md]: Where possible, use "you" when involving actions of the user.  
```suggestion
For each of these customizable Docker commands, a configuration setting is available to set the template of what to execute. Alternatively, you can define multiple templates, optionally with a regular expression which, when matched, hints the context in which a template should be used. The templates support some tokens similar to `launch.json` and `tasks.json`, for example, `${workspaceFolder}`.
```

File: docs/containers/reference.md [hunk 7]
```diff
@@ -268,3 +268,147 @@ Here are all properties available for configuring `docker-run` task. All propert
...
+
+### Settings JSON schema
+
+There are two options for configuring each of the templates (listed below). The first is a single template which overrides the default behavior:
```
Comment (ghogen) [inline docs/containers/reference.md]:
```suggestion
You have two options for configuring each of the templates (listed below). The first is a single template which overrides the default behavior:
```

File: docs/containers/reference.md [hunk 8]
```diff
@@ -268,3 +268,147 @@ Here are all properties available for configuring `docker-run` task. All propert
...
+
+```json
+{
+    "docker.commands.build": "docker build --rm -f \"${dockerfile}\" -t ${tag} \"${context}\""
+}
+```
+
+The second is multiple templates which will be chosen based on the `match` regular expression as well as user input. For example, two templates are given below:
```
Comment (ghogen) [inline docs/containers/reference.md]:
```suggestion
The second is multiple templates which will be chosen based on the `match` regular expression as well as user input. For example, two templates are shown in the following code:
```

File: docs/containers/reference.md (hunk 1 above)
Comment (bwateratmsft) [reply docs/containers/reference.md]: I only just now learned about this suggested change thing, I love it!

**Comment by scscgit** [comment]
**I know this doesn't relate to the docs affected by this pull request**, as they omit examples, but I want to mention for people who may come here that the https://github.com/microsoft/vscode-docker/issues/1596 issue (which is closed and doesn't allow new comments), which is currently a leading google search result for VS Code tail (customization) issues, has a typo of "-tail" instead of "--tail", and that it makes more sense to also combine this with "--follow" to display new output in real time, so I think a better recommended customized value of "docker.commands.logs" (for people annoyed by being unable to view logs of long-running services) would be:

> "docker logs --follow --tail 1000 ${containerId}"

(Personally I think it would be ideal if docs or even the Settings window had some popular & opinionated use-cases covered as a recommendation on top of just providing the default.)

**Comment by bwateratmsft** [comment]
@scscgit I fixed the typo on that GitHub issue, thanks! I think we should consider changing the default of that command to include `--tail`, so I opened https://github.com/microsoft/vscode-docker/issues/2369.

**Pull request merged**